### PR TITLE
Refactor update-ips script

### LINKED ISSUE #1750: Improve the logic of update-ips script
**Is your feature request related to a problem? Please describe.**
With the work of #1582 done we're improved the functions and general handling of IPs a lot, so now we should focus on improving the logic behind it.

**Describe the solution you'd like**
Better grouping of checks so that as an example all specific for SC may be under the same if statement.
Reusable function for looking up keystone+swift endpoints.
Reusable functions for extracting domain name and ports from endpoints.
Simplify the logic for rclone sync.
Moving things to functions to simplify structure and make it more digestible.

**Describe alternatives you've considered**
Keep current which is harder to maintain.

**Definition of done:**
The update-ips script is further improved.

### PR DESCRIPTION
**Pull request by simonklb**
> [!warning]
> **This is public repository, ensure not to disclose:**
> - [ ] personal data beyond what is necessary for interacting with this pull request
> - [ ] business confidential information, such as customer names

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [ ] kind/feature       
- [x] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [ ] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change  
- [ ] kind/dev-change    
- [ ] kind/security      
- [ ] [kind/adr]()       

### What does this PR do / why do we need this PR?

- Part of #1750

#### Additional information to reviewers

Currently only added unit tests for the current state of the script and improved the error handling:
- Some errors didn't exit the script.
- Some errors didn't exit early or failed without a proper message.

I wanted to a fairly extensive test coverage before I start refactoring to not accidentally break the script. However, due to the code complexity I had to do it more or less line by line because it was too hard to parse the functionality in a more cohesive way.

Would appreciate someone taking a look. At least checking the title of the test cases to see if it is covering the script appropriately.

#### Checklist

- [ ] Proper commit message prefix on all commits
  
- Change checks:
  - [ ] The change is transparent
  - [ ] The change is disruptive
  - [ ] The change requires no migration steps
  - [ ] The change requires migration steps
- Metrics checks:
  - [ ] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [ ] The logs do not show any errors after the change
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packages in the `NetworkPolicy Dashboard`
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Audit checks:
  - [ ] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Bug checks:
  - [ ] The bug fix is covered by regression tests

### CODE REVIEW

File: bin/update-ips.bash [hunk 1]
```diff
@@ -282,7 +296,7 @@ else
   DIG_CLUSTER="wc"
 fi
 S3_ENDPOINT="$(yq_dig "${DIG_CLUSTER}" '.objectStorage.s3.regionEndpoint' '""' | sed 's/https\?:\/\///' | sed 's/[:\/].*//')"
-if [[ "${S3_ENDPOINT}" == "" ]]; then
+if [[ "${S3_ENDPOINT}" == "" || "${S3_ENDPOINT}" == '""' ]]; then
```
Comment (Xartos) [inline bin/update-ips.bash:299]: Is this because the line above? `yq_dig "${DIG_CLUSTER}" '.objectStorage.s3.regionEndpoint' '""'`, if so could we change the last parameter to just be `''` to solve it that way instead?

File: bin/update-ips.bash [hunk 2]
```diff
@@ -104,8 +105,9 @@ checkIfDiffAndUpdateKubectlIPs() {
   local config_key="${3}"
   local config_file="${4}"
 
-  local -a ips
-  readarray -t ips <<< "$(getKubectlIPs "${cluster}" "${label}" | tr ' ' '\n')"
+  local -a ips kubectlIPs
+  kubectlIPs=$(getKubectlIPs "${cluster}" "${label}")
+  readarray -t ips <<< "$(echo "${kubectlIPs}" | tr ' ' '\n')"
```
Comment (Xartos) [inline bin/update-ips.bash:110]: TIL:

```console
$ readarray --help
...
    A synonym for `mapfile'.
```

I typically use `mapfile` so I was wondering what the difference bewteen these were

**Review by Xartos — COMMENTED** [review]
Dear god 😱 1090 lines of tests for this script, props to you to for doing this

File: pipeline/resources/update-ips-dry-run-rclone-secondary-add.out [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+[32m+- 127.0.3.10/32[0m
+[32m+- 127.0.3.11/32[0m
+[32m+- 127.0.3.12/32[0m
+[[33mck8s[0m] Diff found for .networkPolicies.global.wcNodes.ips in wc-config.yaml (diff shows actions needed to be up to date)
+[1m--- sc-config.yaml[0m
+[1m+++ expected[0m
+[36m@@ -1 +1 @@[0m
+[31m-[][0m
+[32m+- 127.0.0.4/32[0m
+[[33mck8s[0m] Diff found for .networkPolicies.rcloneSync.secondaryUrl.ips in sc-config.yaml (diff shows actions needed to be up to date)
```
Comment (linus-astrom) [inline pipeline/resources/update-ips-dry-run-rclone-secondary-add.out:98]: I can comment that `secondaryUrl` main purpose is not really to be an endpoint for rclone.
Its intended purpose was if the used url from `s3` or `swift` didn't give all the ips/ports it needed for accessing the endpoint. 
Which happened to me one time it errored out because it couldn't access the next url in line through the specified url because of not allowing the specified ip/port that other url used so by also putting in the second url it could upload the data.
But this way works too :).

File: pipeline/update-ips.bats [hunk 4]
```diff
@@ -0,0 +1,1088 @@
...
+_test_swift_requires_username_or_application_credential_id() {
+  _setup_base
+
+  yq4 -i "${1}"' = "swift"' "${CK8S_CONFIG_PATH}/sc-config.yaml"
+  yq4 -i '.objectStorage.swift.authUrl = "https://keystone.foo.dev-ck8s.com"' "${CK8S_CONFIG_PATH}/sc-config.yaml"
+  yq4 -i '.objectStorage.swift.region = "swift-region"' "${CK8S_CONFIG_PATH}/sc-config.yaml"
+
+  mock_set_output "${mock_dig}" "127.0.0.4" 4 # $os_auth_endpoint
+
+  run ck8s update-ips both dry-run
```
Comment (linus-astrom) [inline pipeline/update-ips.bats:229]: Any reason for using `update-ips both` when you seemingly only use sc-config?

File: bin/update-ips.bash (hunk 1 above)
Comment (simonklb) [reply bin/update-ips.bash]: Absolutely, I thought it was like that for a reason but I don't see why. https://github.com/elastisys/compliantkubernetes-apps/pull/1853/commits/317caca72d6c5df1eb1304b8e240d852d33f1e03

File: pipeline/update-ips.bats (hunk 4 above)
Comment (simonklb) [reply pipeline/update-ips.bats:229]: I decided to only add tests for `update-ips both` as it should hopefully add test coverage for all cases. Do you know if there is any parts of the script that is not evaluated if you run `both`?

File: pipeline/resources/update-ips-dry-run-rclone-secondary-add.out (hunk 3 above)
Comment (linus-astrom) [reply pipeline/resources/update-ips-dry-run-rclone-secondary-add.out:98]: yeah, at the beginning i was just getting this output confused and thought it made secondaryurl into a endpoint.

File: pipeline/update-ips.bats (hunk 4 above)
Comment (linus-astrom) [reply pipeline/update-ips.bats:229]: Everything should be evaluated when you use `both` i was just thinking that it would take longer to go through it all when your using `both` instead of `sc` seeing as most of the script you are only evaluating `sc`.

Comment (simonklb) [reply pipeline/update-ips.bats:229]: Ah I see! It's possible that the test can be optimized once the script is refactored. Since it was completely new to me I just went the safest route possible to begin with. I'll keep it in mind, thanks!

**Comment by simonklb** [comment]
> Would it be possible in this PR to also add support for plain IPs for s3 instead of a domain? With the current script it produces a python exception. There have been some cases on-prem where we only get an IP.

I don't want to introduce any functional changes as part of the refactoring. But I would be fine doing it as a separate change. Do we have an issue for it?

**Comment by Elias-elastisys** [comment]
> I don't want to introduce any functional changes as part of the refactoring. But I would be fine doing it as a separate change. Do we have an issue for it?

No never created one, but I'll do it now.

**Comment by Elias-elastisys** [comment]
Issue here https://github.com/elastisys/compliantkubernetes-apps/issues/1857

**Comment by simonklb** [comment]
One functional change has been made:
The dry-run previously was a bit inconsistent. For example, when diffing lists of IPs it was only showing the lists but for other things the diff of the full config file was used. Now every diff is of the full config file with 3 lines of context. Feel free to chime in if you think it should be different!

File: bin/update-ips.bash [hunk 5]
```diff
@@ -1,113 +1,149 @@
...
 
-# Fetches the Internal IP and calico tunnel ip of kubernetes nodes using the label selector.
-# If label selector isn't specified, all nodes will be returned.
-getKubectlIPs() {
-  local IPS_internal
-  local IPS_calico
-  local IPS
+# Fetches the InternalIP, Calico tunnel IP and Wireguard IP of Kubernetes
+# nodes using the label selector. If label selector isn't specified, all
+# nodes will be returned.
```
Comment (linus-astrom) [inline bin/update-ips.bash:84]:
```suggestion
# Fetches the InternalIP, Calico tunnel IP and Wireguard IP of Kubernetes nodes using the label selector.
# If label selector isn't specified, all nodes will be returned.
```

File: bin/update-ips.bash [hunk 6]
```diff
@@ -146,365 +185,330 @@ process_ips_to_cidrs() {
...
-
-# checkIfDiffAndUpdatePorts <yaml_path> <file_path> <port 1> <port ..>
-checkIfDiffAndUpdatePorts() {
-  yaml_path="${1}"
-  file="${2}"
-  shift 2
-
-  ports="$(echo "[$(for port in "$@"; do echo "$port,"; done)]" | yq4 -oj)"
+# Get the value of the config option or the provided default value if the
+# config option is unset.
```
Comment (linus-astrom) [inline bin/update-ips.bash:189]:
```suggestion
# Get the value of the config option or the provided default value if the config option is unset.
```

**Review by Xartos — APPROVED** [review]

File: bin/update-ips.bash (hunk 5 above)
Comment (simonklb) [reply bin/update-ips.bash:84]: I have a habit of cutting of comments at 80 characters. Do we have any guidelines that says this is incorrect or why are you suggesting this?

Comment (linus-astrom) [reply bin/update-ips.bash:84]: for me it just feels like 80 characters is a bit low to have a whole sentence just cut off at while continuing on the other line  at least with something in between like the `#`.
but otherwise without my feelings in the way it can at least be good to start a new sentence in a new line like this so its clearly marked that it is a new sentence.
```suggestion
# Fetches the InternalIP, Calico tunnel IP and Wireguard IP of Kubernetes
# nodes using the label selector. 
# If label selector isn't specified, all nodes will be returned.
```

**Comment by simonklb** [comment]
@linus-astrom proposed elsewhere to sort IPs which I thought was a good idea. Currently they are not sorted, with this they should be. https://github.com/elastisys/compliantkubernetes-apps/pull/1853/commits/b30859ecde0b3d3f5e990fb3c5b4795a39a11364 Let me know what the rest of you think!

File: bin/update-ips.bash (hunk 5 above)
Comment (simonklb) [reply bin/update-ips.bash]: As long as we don't have a guideline saying otherwise I'll stick to 80 chars. I would not be against having a guideline though.

I think your suggestion to add some spacing to the comments make sense though https://github.com/elastisys/compliantkubernetes-apps/pull/1853/commits/489d45fe3b9443137528947e74c6742dacae2b7d

File: bin/update-ips.bash [hunk 7]
```diff
@@ -143,368 +260,280 @@ process_ips_to_cidrs() {
...
+  swift_enabled || return 0
 
-## Add sc ingress ips to common config
-checkIfDiffAndUpdateDNSIPs "grafana.${OPS_DOMAIN}" ".networkPolicies.global.scIngress.ips" "${config["override_common"]}"
+  local os_auth_url
+  local os_auth_host
+  local os_auth_port
+  os_auth_url="$(yq_read "sc" '.objectStorage.swift.authUrl' "")"
+  os_auth_host=$(parse_url_host "${os_auth_url}")
+  os_auth_port=$(parse_url_port "${os_auth_url}" 5000)
```
Comment (aarnq) [inline bin/update-ips.bash:391]: Shouldn't the default port always be based on the protocol?

**Review by aarnq — APPROVED** [review]
Huge improvement!

File: bin/update-ips.bash (hunk 7 above)
Comment (simonklb) [reply bin/update-ips.bash:391]: That makes sense to me but with this PR my goal is only to refactor the code, not make any behavior changes. I realize I did stray from that goal when sorting IPs but I felt that was fine since the actual result is still going to be the same.

I'll add it to my list of follow-ups, together with https://github.com/elastisys/compliantkubernetes-apps/issues/1857 and your offline remark on `.networkPolicies.rcloneSync.destinationObjectStorageSwift` maintaining all the necessary config and not relying on `.networkPolicies.global.objectStorage`.

**Review by linus-astrom — APPROVED** [review]

**Comment by simonklb** [comment]
Would love it if you could dry-run on a few of your environments to verify that nothing diffs (except for maybe the IPs being reordered due to the new sorting).

**Comment by aarnq** [comment]
Tested it on a few environments and it works well, including Swift + Rclone :+1:

File: bin/update-ips.bash (hunk 7 above)
Comment (simonklb) [reply bin/update-ips.bash:391]: For posterity:

> Shouldn't the default port always be based on the protocol?

https://github.com/elastisys/compliantkubernetes-apps/pull/1876

> and your offline remark on `.networkPolicies.rcloneSync.destinationObjectStorageSwift` maintaining all the necessary config and not relying on `.networkPolicies.global.objectStorage`.

https://github.com/elastisys/compliantkubernetes-apps/pull/1873
